'''
load apriori and association modules from mlxtend.frequent_patterns 
Used different dataset because mlxtend need data in below format. 

    transaction_name    apple banana grapes
transaction  1            0     1      1
             2            1     0      1  
             3            1     0      0
             4            0     1      0
             
 we could have used above data as well but need to perform operation to bring in this format instead of that used seperate data only.  

In [ ]:
#pip install mlxtend

In [6]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
df= pd.read_csv("C:\\Users\\hp\\Desktop\\apriopri\\data-2.csv" ,encoding="ISO-8859-1" )

In [9]:
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [24]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [14]:
df1= df[df.Country== "France"]
df1["Description"]= df1["Description"].str.strip()

C:\Users\hp\AppData\Local\Temp\ipykernel_7228\835054084.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1["Description"]= df1["Description"].str.strip()


In [26]:
df1= df1[df1.Quantity>0]   #some negative values
basket= pd.pivot_table(data= df1, index= "InvoiceNo", columns="Description", values="Quantity", aggfunc="sum", fill_value=0)

In [28]:
basket.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                            0  ...                                0
536852                            0  ...                                0
536974                            0  ...                                0
537065                            0  ...                                0
537463                            0  ...                                0

[5 rows x 1563 columns]

In [32]:
#no need of quantity sum rather need yes or no buy
def convert_binary(item):
    if item>0: 
        return 1
    else :
        return 0

In [34]:
basket_sets= basket.applymap(convert_binary)       # to each item row and columns applymap

C:\Users\hp\AppData\Local\Temp\ipykernel_7228\4039380815.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_sets= basket.applymap(convert_binary)


In [45]:
#basket_sets.drop(columns=['POSTAGE'],inplace=True)

In [46]:
# call apriori function and pass minimum support here we are passing 7%. 
# means 7 times in total number of transaction the item should be present.

frequent_itemsets= apriori(basket_sets, min_support=0.05, use_colnames=True)

In [47]:
frequent_itemsets

support                                           itemsets
0    0.071429                      (4 TRADITIONAL SPINNING TOPS)
1    0.096939                       (ALARM CLOCK BAKELIKE GREEN)
2    0.102041                        (ALARM CLOCK BAKELIKE PINK)
3    0.094388                         (ALARM CLOCK BAKELIKE RED)
4    0.068878                       (ASSORTED COLOUR MINI CASES)
..        ...                                                ...
108  0.102041  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...
109  0.122449  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...
110  0.063776  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...
111  0.068878  (PLASTERS IN TIN SPACEBOY, PLASTERS IN TIN CIR...
112  0.099490  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...

[113 rows x 2 columns]

Frequent itemsets are the ones which occur at least a minimum number of times in the transactions. Technically, these are the itemsets for which support value (fraction of transactions containing the itemset) is above a minimum threshold — min_support. We have kept min_support=0.05 in our above code.

So, {Bread, Notebook} might not be a frequent itemset if it occurs only 2 times out of 100 transactions and (2/100) = 0.02 falls below the value of min_support.

In [49]:
#applying association rules on frequent itemset

rules_mlxtend= association_rules(frequent_itemsets, metric= "lift", min_threshold=1)
rules_mlxtend

antecedents  ... zhangs_metric
0                         (ALARM CLOCK BAKELIKE PINK)  ...      0.964734
1                        (ALARM CLOCK BAKELIKE GREEN)  ...      0.959283
2                        (ALARM CLOCK BAKELIKE GREEN)  ...      0.979224
3                          (ALARM CLOCK BAKELIKE RED)  ...      0.976465
4                         (ALARM CLOCK BAKELIKE PINK)  ...      0.968652
..                                                ...  ...           ...
81  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...  ...      0.967949
82  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...  ...      0.956294
83                      (SET/6 RED SPOTTY PAPER CUPS)  ...      0.995904
84                    (SET/6 RED SPOTTY PAPER PLATES)  ...      0.996251
85               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...      0.964706

[86 rows x 10 columns]

Lift is a measure of how much more likely two items are to be bought together than if they were bought independently. A higher lift indicates a stronger association.

Confidence is a measure of how often the rule has been found to be true. A confidence of 0.8 means the rule is true 80% of the time.



In [52]:
# rules_mlxtend.rename(columns={'antecedents':'lhs','consequents':'rhs'})

# as based business use case we can sort based on confidance and lift.

rules_mlxtend[(rules_mlxtend["lift"]>=4) & (rules_mlxtend["confidence"]>=0.8 )] 

antecedents  ... zhangs_metric
2                        (ALARM CLOCK BAKELIKE GREEN)  ...      0.979224
3                          (ALARM CLOCK BAKELIKE RED)  ...      0.976465
8                        (CHILDRENS CUTLERY SPACEBOY)  ...      0.991123
9                      (CHILDRENS CUTLERY DOLLY GIRL)  ...      0.993846
38                       (PACK OF 6 SKULL PAPER CUPS)  ...      0.993188
39                     (PACK OF 6 SKULL PAPER PLATES)  ...      0.985135
64                    (SET/6 RED SPOTTY PAPER PLATES)  ...      0.956140
66                      (SET/6 RED SPOTTY PAPER CUPS)  ...      0.993343
67                    (SET/6 RED SPOTTY PAPER PLATES)  ...      0.981725
68  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...  ...      0.961653
69  (ALARM CLOCK BAKELIKE PINK, ALARM CLOCK BAKELI...  ...      0.958457
70  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...  ...      0.948476
80  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...  ...      0.953488
81  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...  ...      0.967949
82  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...  ...      0.956294

[15 rows x 10 columns]

antecedents and consequents -> The IF component of an association rule is known as the antecedent. The THEN component is known as the consequent. The antecedent and the consequent are disjoint; they have no items in common.

antecedent support -> This measure gives an idea of how frequent antecedent is in all the transactions. Like (ALARM CLOCK BAKELIKE RED) is present in 9.4% of the transactions.

consequent support -> This measure gives an idea of how frequent consequent is in all the transactions. Like (ALARM CLOCK BAKELIKE GREEN) is present in 9.6% of the transactions.

support -> This measure gives an idea of how frequent ItemSet is in all the transactions. Like {ALARM CLOCK BAKELIKE RED,ALARM CLOCK BAKELIKE GREEN} is present in 7.9% of the transactions.

confidance -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedents. So {ALARM CLOCK BAKELIKE RED} -> {ALARM CLOCK BAKELIKE GREEN} as a confidence of 83%. In simple words their is an 83% chance of finding {ALARM CLOCK BAKELIKE GREEN} , if the cart contains {ALARM CLOCK BAKELIKE RED}.

lift -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedent, but controlling the popularity of consequent. So lift of {ALARM CLOCK BAKELIKE GREEN} w.r.t {ALARM CLOCK BAKELIKE RED} is 8.64. Which is quite good. Any lift value > 1 implies that the Association rule is worth considering. [ Reference : https://en.wikipedia.org/wiki/Lift_(data_mining)]

leverage -> leverage(X -> Y) = P(X and Y) - (P(X)P(Y))

Leverage measures the difference of X and Y appearing together in the data set and what would be expected if X and Y where statistically dependent. The rational in a sales setting is to find out how many more units (items X and Y together) are sold than expected from the independent sells. 